src/libraries/VanityAddressLib.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+/// @title VanityAddressLib
+/// @notice A library to score addresses based on their vanity
+library VanityAddressLib {
+    /// @notice Compares two addresses and returns true if the first address has a better vanity score
+    /// @param first The first address to compare
+    /// @param second The second address to compare
+    /// @return better True if the first address has a better vanity score
+    function betterThan(address first, address second) internal pure returns (bool better) {
+        return score(first) > score(second);
+    }
+
+    /// @notice Scores an address based on its vanity
+    /// @param addr The address to score
+    /// @return calculatedScore The vanity score of the address
+    function score(address addr) internal pure returns (uint256 calculatedScore) {
+        //  Scoring rules:
+        //     Requirement: The first nonzero nibble must be 4
+        //     10 points for every leading 0 nibble
+        //     40 points if the first 4 is followed by 3 more 4s
+        //     20 points if the first nibble after the 4 4s is NOT a 4
+        //     20 points if the last 4 nibbles are 4s
+        //     1 point for every 4
+
+        // convert the address to bytes for easier parsing
+        bytes20 addrBytes = bytes20(addr);
+
+        // 10 points per leading zero nibble
+        uint256 leadingZeroCount = getLeadingNibbleCount(addrBytes, 0, 0);
```

Comment:
Hacker's delight has a binary search that does bring in some gas savings (example below tested but not reviewed)

```solidity
function hdGetLeadingNibbleCount(bytes20 addrBytes) internal pure returns (uint256) {
    uint256 n;
    bytes20 x = addrBytes;
    // assume addrBytes validated to be non zero elsewhere
    assembly {
        x := shr(96, x)
        if iszero(shr(128, x)) {
            n := add(n, 32)
            x := shl(128, x)
        }
        if iszero(shr(192, x)) {
            n := add(n, 16)
            x := shl(64, x)
        }
        if iszero(shr(224, x)) {
            n := add(n, 8)
            x := shl(32, x)
        }
        if iszero(shr(240, x)) {
            n := add(n, 4)
            x := shl(16, x)
        }
        if iszero(shr(248, x)) {
            n := add(n, 2)
            x := shl(8, x)
        }
        if iszero(shr(249, x)) {
            n := add(n, 1)
            x := shl(4, x)
        }
    }
    return n - 24;
}
```